src/ui.jsx

```diff
@@ -0,0 +1,392 @@
+                cols: 0
+            },
+
+            outputMessages: []
+
+        };
+
+        this.setIntervalId = -1;
+
+        // TODO: Used for debugging, push state to this array after every tick so we can examine it later
+        this.savedStates = [];
+
+    }
+
+    /**
+     * Set the initial state of the world.
+     * @param  {String} instructionsString The user's input taken straight from the editor.
+     */
+    _setup(instructionsString) {
+
+        var instructionStack;
+
+        // Convert the instructions passed in by the user into a stack
+        try {
+            instructionStack = parseInstructions(instructionsString);
+        } catch (error) {
+            console.log(error);
+            return;
+        }
+
+        // The first item on the stack is the planet's boundaries
+        var planetBoundaries = instructionStack.pop().trim().split(" ");
+
+        // Use the instruction stack to set the initial state of the world
+        var initialRobots = prepareRobots(instructionStack, planetBoundaries);
+        var initialPlanet = { scents: {}, rows: parseInt(planetBoundaries[0], 10), cols: parseInt(planetBoundaries[1], 10) };
+
+        // Update the state of the world accordingly: The callback starts the simulation
+        this.setState({ robots: initialRobots, planet: initialPlanet },
+
+            () => {
+                this.setIntervalId = setInterval(this._tick.bind(this), 100);
+            }
+
+        );
+
+    }
+
+    // Update everything in the world by one instruction
+    // TODO: Make all of the data accesses immutable
+    _tick() {
+
+        // Use executeInstruction once on each robot to get the next state of the world
+        var planet = this.state.planet;
+        var robots = this.state.robots;
+        var outputMessages = this.state.outputMessages.slice(0);
+        var doneCount = 0;
+
+        // Map is immutable, return result to new array
+        var newRobots = robots.map( robot => {
+
+                // If the robot does not have any instructions to execute or is lost, bail
+                if (robot.lost || robot.instructions.length === 0) {
+                    doneCount++;
+                    return robot;
+                }
+
+                // Get a new robot by updating the state of the current one being processed
+                var newRobot = executeInstruction(robot, planet, robot.instructions[robot.instructions.length - 1]);
+                newRobot.instructions.pop();
+
+                if (newRobot.lost) {
+                    // Leave a scent
+                    this.setState({ planet: { scents: { [`${newRobot.x},${newRobot.y}`]: true } } } );
+                }
+
+                return newRobot;
+
+            }
+
+        );
+
+        // TODO: Debugging, push the current state so we can examine it later
+        this.savedStates.push(this.state);
+
+        // If every robot is done, we're finished
+        if (doneCount === robots.length) {
```

Comment:
You could have said, if all robots have an empty set of instructions, you're done?